### DISCUSSION TITLE
Add texture coordinate array, color coordinate array the right way for it to work.

### DISCUSSION THREAD

**Discussion by johnspeny** [axmolengine/axmol#749] [Q&A]
I am using latest adxe version before it changed to axis.

I can't figure out a way to properly add texture coordinates in an array on a triangle. 
[desired image 1](https://learnopengl.com/img/getting-started/textures.png)
[desired image 2(optional though)](https://learnopengl.com/img/getting-started/shaders3.png) 

I seem to be creating 3 lines of _customCommand.createVertexBuffer and _customCommand.updateVertexBuffer, I don't know if that's the right way of doing it.

I also think there is an issue with _customCommand.setVertexDrawInfo(0, m_vertexDataCount); what value should m_vertexDataCount be?
``` cpp
void SampleScene::generateTriangle()
{
	const char* vertexShaderSource = "#version 330 core\n"
		"layout (location = 0) in vec2 a_position;\n"
		"layout (location = 1) in vec2 a_texCoord;\n"
		"layout (location = 2) in vec4 a_color;\n"
		"out vec4 ourColor;\n"
		"out vec2 TexCoord;\n"
		"uniform mat4 u_MVPMatrix;\n"
		"void main()\n"
		"{\n"
		"   gl_Position = u_MVPMatrix * vec4(a_position, 0.0, 1.0);\n"
		"   ourColor = a_color;\n"
		"   TexCoord = a_texCoord;\n"
		"}\0";

	const char* fragmentShaderSource = "#version 330 core\n"
		"out vec4 FragColor;\n"
		"in vec4 ourColor;\n"
		"in vec2 TexCoord;\n"
		"uniform sampler2D u_tex0;\n"
		"void main()\n"
		"{\n"
		"	FragColor = texture(u_tex0, TexCoord);\n"
		"}\0";

	auto program = cocos2d::backend::Device::getInstance()->newProgram(vertexShaderSource, fragmentShaderSource);

	auto& pipelineDescriptor = _customCommand.getPipelineDescriptor();
	_programState = new (std::nothrow) backend::ProgramState(program);
	pipelineDescriptor.programState = _programState;

	auto& vertexLayout = pipelineDescriptor.programState->getVertexLayout();
	const auto& attributeInfo = _programState->getProgram()->getActiveAttributes();
	auto iter = attributeInfo.find("a_position");
	if (iter != attributeInfo.end())
	{
		vertexLayout->setAttribute("a_position",
			iter->second.location,
			backend::VertexFormat::FLOAT2,
			0,
			false);
	}

	iter = attributeInfo.find("a_texCoord");
	if (iter != attributeInfo.end())
	{
		vertexLayout->setAttribute("a_texCoord",
			iter->second.location,
			backend::VertexFormat::FLOAT2,
			2 * sizeof(float),
			false);
	}

	iter = attributeInfo.find("a_color");
	if (iter != attributeInfo.end())
	{
		vertexLayout->setAttribute("a_color",
			iter->second.location,
			backend::VertexFormat::UBYTE4,
			4 * sizeof(float),
			true);
	}

	// stride = 8 * sizeof(float)
	vertexLayout->setLayout(8 * sizeof(float));

	/*
	@brief get uniforms from program state
	*/
	_mvpMatrixLocation = pipelineDescriptor.programState->getUniformLocation("u_MVPMatrix");
	_textureLocation = pipelineDescriptor.programState->getUniformLocation("u_tex0");

	/*
	@brief vertex data
	*/
	float pos_Vertices[] = {
		// positions
		90.0f, 30.0f,
		120.0f, 60.0f,
		150.0f, 30.0f
	};

	float color_Vertices[] = {
		// color
		1.0f, 0.0f, 0.0f, 1.0f,
		0.0f, 1.0f, 0.0f, 1.0f,
		0.0f, 0.0f, 1.0f, 1.0f
	};

	float texture_Vertices[] = {
		// texture
		90.0f, 30.0f,
		120.0f, 60.0f,
		150.0f, 30.0f
	};

	size_t pCount = (sizeof(pos_Vertices) / (sizeof(float) * 2));
	size_t colorCount = (sizeof(color_Vertices) / (sizeof(float) * 4));
	size_t textureCount = (sizeof(texture_Vertices) / (sizeof(float) * 2));

	// total vertex count from the vertex buffer ( 3 vertices for my case)
	m_vertexDataCount = 3;

	// glVertexAttribPointer part
	_customCommand.createVertexBuffer(sizeof(pos_Vertices), pCount, CustomCommand::BufferUsage::DYNAMIC);
	_customCommand.createVertexBuffer(sizeof(color_Vertices), colorCount, CustomCommand::BufferUsage::DYNAMIC);
	_customCommand.createVertexBuffer(sizeof(texture_Vertices), textureCount, CustomCommand::BufferUsage::DYNAMIC);

	/*
	@brief glBufferData(data, size)
	*/
	_customCommand.updateVertexBuffer(pos_Vertices, sizeof(pos_Vertices));
	_customCommand.updateVertexBuffer(color_Vertices, sizeof(color_Vertices));
	_customCommand.updateVertexBuffer(texture_Vertices, sizeof(texture_Vertices));

	/*
	@brief update uniforms
	*/
	const auto& projectionMat = Director::getInstance()->getMatrix(MATRIX_STACK_TYPE::MATRIX_STACK_PROJECTION);
	_customCommand.getPipelineDescriptor().programState->setUniform(_mvpMatrixLocation, projectionMat.m, sizeof(projectionMat.m));

	/*
	@brief texture rendering on shape
	*/
	auto _texture = Director::getInstance()->getTextureCache()->addImage("lowres-desert-ground.png");
	_customCommand.getPipelineDescriptor().programState->setUniform(_textureLocation, &_texture, sizeof(_texture));
	_customCommand.getPipelineDescriptor().programState->setTexture(_textureLocation, 0, _texture->getBackendTexture());

	/*
	@brief: draw type
	*/
	_customCommand.setDrawType(CustomCommand::DrawType::ARRAY);
	_customCommand.setPrimitiveType(CustomCommand::PrimitiveType::TRIANGLE);
}
```

``` cpp
void SampleScene::draw(cocos2d::Renderer* renderer, const cocos2d::Mat4& transform, uint32_t flags)
{
	_customCommand.init(0);
	_customCommand.setVertexDrawInfo(0, m_vertexDataCount);
	renderer->addCommand(&_customCommand);
}
```

The code above runs but with no visible triangle.

Also I know I could just use 1 vertex array like below but I don't want that approach.
```
float vertices[] = {
		// positions              // color                                    // texture
		90.0f, 30.0f,		1.0f, 0.0f, 0.0f, 1.0f,                  90.0f, 30.0f,
		120.0f, 60.0f,		0.0f, 1.0f, 0.0f, 1.0f,			120.0f, 60.0f,
		150.0f, 30.0f,		0.0f, 0.0f, 1.0f, 1.0f,			150.0f, 30.0f
	};
```

**Comment by DelinWorks** [comment]
If you're on windows, are you using `ANGLE` as your backend renderer?

**Reply by johnspeny** [reply]
I am using windows 10, in that file i include something like this. 
#include "renderer/backend/Device.h"

I use opengl es, just the default one.

**Reply by DelinWorks** [reply]
Based on my testing, when `createVertexBuffer` is called. it expects an array like this:

```
float vertices[] = {
		// positions              // color                                    // texture
		90.0f, 30.0f,		1.0f, 0.0f, 0.0f, 1.0f,                  90.0f, 30.0f,
		120.0f, 60.0f,		0.0f, 1.0f, 0.0f, 1.0f,			120.0f, 60.0f,
		150.0f, 30.0f,		0.0f, 0.0f, 1.0f, 1.0f,			150.0f, 30.0f
	};
```

Not a separate array for every vertex attribute like you did in your example. So when it's called it should be called once because it sets up a single internal vertex buffer which if you call the function again will be overridden

![image](https://user-images.githubusercontent.com/45469625/179216936-7ea1fca5-9e15-4567-b834-d2416ce9468a.png)

So I think your only solution is to stuff the vertex data into a single array, what makes you separate arrays like this anyways? isn't the way I mentioned relevant? A single array is much easier to deal with since you can get any data you want and modify it using offsetting:

Let's say I want the third (index 2) vertex's position color and texCoord. I could simply say
```
int perVertexSize = 8;
int index = 2 * perVertexSize;

arr[index + 0] // x
arr[index + 1] // y

arr[index + 2] // r
arr[index + 3] // g
arr[index + 4] // b
arr[index + 5] // a

arr[index + 6] // u
arr[index + 7] // v
``` 

and if the custom command was to support this feature how would it know which vertex buffer you're referring to on these three lines?

```
_customCommand.updateVertexBuffer(pos_Vertices, sizeof(pos_Vertices));
_customCommand.updateVertexBuffer(color_Vertices, sizeof(color_Vertices));
_customCommand.updateVertexBuffer(texture_Vertices, sizeof(texture_Vertices));
```

You should also consider multiplying by the data type size when updating a vertex buffer because GPUs only understand bytes:
`_customCommand.updateVertexBuffer(pos_Vertices, sizeof(pos_Vertices) * sizeof(float));`
or
`_customCommand.updateVertexBuffer(pos_Vertices, sizeof(pos_Vertices) * sizeof(pos_Vertices[0]));`

**Reply by johnspeny** [reply]
I was following a very old tutorial [here](https://chowdera.com/2021/09/20210902034412670v.html), I wanted to store my coordinates using their approach which is using different arrays.

They have these 2 lines that I think are different from your approach or mine. 
```
glVertexAttribPointer(kCCVertexAttrib_Position,  2, GL_FLOAT, GL_FALSE,  0, _hillVertices);
glVertexAttribPointer(kCCVertexAttrib_TexCoords,  2, GL_FLOAT, GL_FALSE,  0, _hillTexCoords)
```

I think am stuck if there is only one approach to it.

**Reply by DelinWorks** [reply]
don't worry it's really easy, most game engines achieve it with using the approach I mentioned, I'll show a more modern solution

**Reply by DelinWorks** [reply]
Also the approach they are doing is not really practical because they seem to access the OpenGL API directly, approved by the `glVertexAttribPointer`, maybe because the version of the engine they're explaining on is old as time

**Reply by DelinWorks** [reply]
This is the right way to do it:
using `QUAD_TEXTURE` you can have a position (3D) and a color and texture coordinates 
```
auto renderer = MeshRenderer::create();

int perVertexSizeInFloat = 9;  // 3+4+2

IndexArray indices;
std::vector<float> vertices;
indices.clear(CustomCommand::IndexFormat::U_INT);

auto InsertVertex = [perVertexSizeInFloat, &vertices, &indices](float x, float y, float z, Color4F col, float u, float v)
{
	unsigned int startindex = vertices.size() / perVertexSizeInFloat;
	vertices.insert(vertices.end(), {
		x,y,z,  col.r,col.g,col.b,col.a, u,v,
		});

	indices.insert<uint32_t>(indices.size(),
		ilist_u32_t{ startindex });
};

InsertVertex(-100, -100, 0, { 1,0,0,1 }, 0, 0);
InsertVertex(100, -100, 0, { 0,1,0,1 }, 1, 0);
InsertVertex(0, 100, 0, { 0,0,1,1 }, .5, 1);

std::vector<MeshVertexAttrib> attribs;
MeshVertexAttrib att;

att.type = backend::VertexFormat::FLOAT3;
att.vertexAttrib = shaderinfos::VertexKey::VERTEX_ATTRIB_POSITION;
attribs.push_back(att);

att.type = backend::VertexFormat::FLOAT4;
att.vertexAttrib = shaderinfos::VertexKey::VERTEX_ATTRIB_COLOR;
attribs.push_back(att);

att.type = backend::VertexFormat::FLOAT2;
att.vertexAttrib = shaderinfos::VertexKey::VERTEX_ATTRIB_TEX_COORD;
attribs.push_back(att);

auto mesh = Mesh::create(vertices, perVertexSizeInFloat, indices, attribs);

auto mat = MeshMaterial::createBuiltInMaterial(MeshMaterial::MaterialType::QUAD_TEXTURE, false);
mesh->setMaterial(mat);
mesh->setTexture("HelloWorld.png");

renderer->addMesh(mesh);

addChild(renderer);

```
If you just want color you have to use `QUAD_COLOR` material and remove these lines
```
att.type = backend::VertexFormat::FLOAT2;
att.vertexAttrib = shaderinfos::VertexKey::VERTEX_ATTRIB_TEX_COORD;
attribs.push_back(att);
```
and remove the texCoords from the vertex list and change perVertexSizeInFloat to reflect the new vertex size (which is 7)

You should see a triangle appear in the bottom left just set the position of the camera to 0,0 to see the triangle in the middle

![image](https://user-images.githubusercontent.com/45469625/179230380-3a8f2811-b7d2-4f86-a620-73834e949182.png)

![image](https://user-images.githubusercontent.com/45469625/179231559-a6792ca8-7bbc-40a3-ab14-ea91e6ec227b.png)

**Reply by DelinWorks** [reply]
If you want to update the vertices on runtime you should first acquire the GPU buffer that handles the vertices
You could use `mesh->getMeshIndexData()->getVertexBuffer()`

If anything confuses you feel free to ask!

**Reply by johnspeny** [reply]
I am trying out your example but I can not seem to find MeshRenderer in adxe or even cocos2dx v4 docs
```
auto renderer = MeshRenderer::create();
```

**Reply by DelinWorks** [reply]
The `MeshRenderer` class is a new renamed type that was merged in dev branch, you have to pull to the latest dev commit

**Reply by johnspeny** [reply]
This is my code snippet for color triangle without texture. Texture example you sent works well.
``` cpp
	auto renderer = MeshRenderer::create();

	int perVertexSizeInFloat = 7;  // 3+4

	IndexArray indices;
	std::vector<float> vertices;
	indices.clear(CustomCommand::IndexFormat::U_INT);

	auto InsertVertex = [perVertexSizeInFloat, &vertices, &indices](float x, float y, float z, Color4F col)
	{
		unsigned int startindex = vertices.size() / perVertexSizeInFloat;
		vertices.insert(vertices.end(), {
			x,y,z,  col.r,col.g,col.b,col.a
			});

		indices.insert<uint32_t>(indices.size(),
			ilist_u32_t{ startindex });
	};

	InsertVertex(-100, -100, 0, { 1,0,0,1 });
	InsertVertex(100, -100, 0, { 0,1,0,1 });
	InsertVertex(0, 100, 0, { 0,0,1,1 });

	std::vector<MeshVertexAttrib> attribs;
	MeshVertexAttrib att;

	att.type = backend::VertexFormat::FLOAT3;
	att.vertexAttrib = shaderinfos::VertexKey::VERTEX_ATTRIB_POSITION;
	attribs.push_back(att);

	att.type = backend::VertexFormat::FLOAT4;
	att.vertexAttrib = shaderinfos::VertexKey::VERTEX_ATTRIB_COLOR;
	attribs.push_back(att);

	auto mesh = Mesh::create(vertices, perVertexSizeInFloat, indices, attribs);

	auto mat = MeshMaterial::createBuiltInMaterial(MeshMaterial::MaterialType::QUAD_COLOR, false);
	mesh->setMaterial(mat);

	renderer->addMesh(mesh);

	addChild(renderer);
```

The program crashes whenever it reaches mesh->setMaterial(mat). I do not know what I am doing wrong.

**Reply by DelinWorks** [reply]
It works fine on my end! that's weird...

![image](https://user-images.githubusercontent.com/45469625/179667845-6dd2438d-2495-41f4-a463-cde50dc80701.png)

that's the code I used in case there are some letters that are different than yours
```
auto renderer = MeshRenderer::create();

int perVertexSizeInFloat = 7;  // 3+4

IndexArray indices;
std::vector<float> vertices;
indices.clear(CustomCommand::IndexFormat::U_INT);

auto InsertVertex = [perVertexSizeInFloat, &vertices, &indices](float x, float y, float z, Color4F col)
{
	unsigned int startindex = vertices.size() / perVertexSizeInFloat;
	vertices.insert(vertices.end(), {
		x,y,z,  col.r,col.g,col.b,col.a,
		});

	indices.insert<uint32_t>(indices.size(),
		ilist_u32_t{ startindex });
};

InsertVertex(-200, -100, 0, { 1,0,0,1 });
InsertVertex(200, -100, 0, { 0,1,0,1 });
InsertVertex(0, 200, 0, { 0,0,1,1 });

std::vector<MeshVertexAttrib> attribs;
MeshVertexAttrib att;

att.type = backend::VertexFormat::FLOAT3;
att.vertexAttrib = shaderinfos::VertexKey::VERTEX_ATTRIB_POSITION;
attribs.push_back(att);

att.type = backend::VertexFormat::FLOAT4;
att.vertexAttrib = shaderinfos::VertexKey::VERTEX_ATTRIB_COLOR;
attribs.push_back(att);

auto mesh = Mesh::create(vertices, perVertexSizeInFloat, indices, attribs);

auto mat = MeshMaterial::createBuiltInMaterial(MeshMaterial::MaterialType::QUAD_COLOR, false);
mesh->setMaterial(mat);

renderer->addMesh(mesh);
```

What error message are you getting?
Maybe the `QUAD_COLOR` shader has a compilation error on your GPU which makes the program state null thus crashing.

Mine is a `GeForce GTX 1060` what's yours?

**Reply by DelinWorks** [reply]
Oh now I know!
it's a vendor specific error that's why it only works for me

in the vertex shader
```
attribute vec2 a_texCoord;
```
this line under the vertex shader of quad expects a vertex attribute for texture coordinates, when you want to render a color-only vertex the shader fails to compile because the fragment shader is not dealing with this attribute.. or the vertex shader isn't getting vertex sizes based on the total attributes, that's why the shader fails to compile! I'll do a PR to fix this.

#752

**Reply by johnspeny** [reply]
I got it working well for both triangles i.e colored and textured triangles thanks.
I also added the camera to center the triangles, I hope that's the modern way of doing it. 

Out of curiosity, on [image](https://user-images.githubusercontent.com/45469625/179667845-6dd2438d-2495-41f4-a463-cde50dc80701.png) around the top left corner. Is that stats display part of Axis?. How do I obtain it?.

Working code snippet
``` cpp
// Camera
	auto s = Director::getInstance()->getWinSize();
	auto camera = Camera::createOrthographic(s.width, s.height, 1, 1000);
	camera->setCameraFlag(CameraFlag::USER1);
	camera->setPosition3D(Vec3(0, 0, 1));
	addChild(camera);

	// Triangle
	auto renderer = MeshRenderer::create();

	int perVertexSizeInFloat = 7;  // 3+4

	IndexArray indices;
	std::vector<float> vertices;
	indices.clear(CustomCommand::IndexFormat::U_INT);

	auto InsertVertex = [perVertexSizeInFloat, &vertices, &indices](float x, float y, float z, Color4F col)
	{
		unsigned int startindex = vertices.size() / perVertexSizeInFloat;
		vertices.insert(vertices.end(), {
			x,y,z,  col.r,col.g,col.b,col.a
			});

		indices.insert<uint32_t>(indices.size(),
			ilist_u32_t{ startindex });
	};

	InsertVertex(-100, -100, 0, { 1,0,0,1 });
	InsertVertex(100, -100, 0, { 0,1,0,1 });
	InsertVertex(0, 100, 0, { 0,0,1,1 });

	std::vector<MeshVertexAttrib> attribs;
	MeshVertexAttrib att;

	att.type = backend::VertexFormat::FLOAT3;
	att.vertexAttrib = shaderinfos::VertexKey::VERTEX_ATTRIB_POSITION;
	attribs.push_back(att);

	att.type = backend::VertexFormat::FLOAT4;
	att.vertexAttrib = shaderinfos::VertexKey::VERTEX_ATTRIB_COLOR;
	attribs.push_back(att);

	auto mesh = Mesh::create(vertices, perVertexSizeInFloat, indices, attribs);

	auto mat = MeshMaterial::createBuiltInMaterial(MeshMaterial::MaterialType::QUAD_COLOR, false);
	mesh->setMaterial(mat);

	renderer->addMesh(mesh);

	addChild(renderer);

	renderer->setCameraMask(2);
```

**Reply by DelinWorks** [reply]
Its better to use the default camera of the scene rather than creating your own. you can obtain it with `_defaultCamera` or `Camera::getDefaultCamera();`

You shouldn't use the getDefaultCamera on scene init because that's a guaranteed crash. use `_defaultCamera` private member of scene directly.

as for the stats top left that's not axis that's an external thirdparty software that attaches itself to directx and opengl contexts and it shows stats, it's called `MSI Afterburner`

**Reply by johnspeny** [reply]
How would a triangle strip code snippet look like using this approach. Is the DrawType a triangle by default?

**Reply by DelinWorks** [reply]
you would need to change the primitive draw type of the custom command in the mesh renderer source file, and yes the draw type is triangle by default, i think it would be nice to make a set function in the mesh renderer class.

**Reply by johnspeny** [reply]
![Untitled](https://user-images.githubusercontent.com/30299284/179830674-0e09f91f-3d44-4c5f-87b3-cb09ffaced43.png)

Using the current approach that is what I obtained. The old way had primitive type as triangle or triangle strip. I probably want the triangle strip. Hope that helps if I wasn't clear on the issue earlier on.

**Reply by DelinWorks** [reply]
don't be afraid of using the current system as it's just an abstraction over the renderer class and the renderer class is another abstraction over opengl and that makes it easy to add any feature you like. don't worry I'll add custom primitives in my free time.

**Reply by johnspeny** [reply]
I am using the current system, it's cool plus it will help others in the future. So far it's good as improvements are being made as well.

**Reply by DelinWorks** [reply]
I've been thinking about it and design-wise the material is the one that should store the primitive draw type not the mesh because if instancing were to be implemented, then different mesh transforms cannot have different draw types, the material is the one that should decide their draw type, it seems to be neater this way.

**Reply by DelinWorks** [reply]
With PR #757 you can now specify a primitive type in the material like so:

`mat->setPrimitiveType(backend::PrimitiveType::TRIANGLE);`

![image](https://user-images.githubusercontent.com/45469625/179865137-ae6441b3-9fd6-490c-8099-e30f640fd10c.png)

Or in your case `mat->setPrimitiveType(backend::PrimitiveType::TRIANGLE_STRIP);`

![image](https://user-images.githubusercontent.com/45469625/179865347-a37005b4-ec6a-4cf9-8023-6f03821aa919.png)


This is the code
```
auto renderer = MeshRenderer::create();

int perVertexSizeInFloat = 7;  // 3+4+2

IndexArray indices;
std::vector<float> vertices;
indices.clear(CustomCommand::IndexFormat::U_INT);

auto InsertVertex = [perVertexSizeInFloat, &vertices, &indices](float x, float y, float z, Color4F col)
{
	unsigned int startindex = vertices.size() / perVertexSizeInFloat;
	vertices.insert(vertices.end(), {
		x,y,z,  col.r,col.g,col.b,col.a,
		});

	indices.insert<uint32_t>(indices.size(),
		ilist_u32_t{ startindex });
};

float offset = -100;

InsertVertex(-200 + offset * 2, -100 + offset, 0, { 1,0,0,1 });
InsertVertex(200 + offset * 2, -100 + offset, 0, { 0,1,0,1 });
InsertVertex(0 + offset * 2, 200 + offset, 0, { 0,0,1,1 });

offset = 100;

InsertVertex(-200 + offset * 2, -100 + offset, 0, { 1,0,0,1 });
InsertVertex(200 + offset * 2, -100 + offset, 0, { 0,1,0,1 });
InsertVertex(0 + offset * 2, 200 + offset, 0, { 0,0,1,1 });

std::vector<MeshVertexAttrib> attribs;
MeshVertexAttrib att;

att.type = backend::VertexFormat::FLOAT3;
att.vertexAttrib = shaderinfos::VertexKey::VERTEX_ATTRIB_POSITION;
attribs.push_back(att);

att.type = backend::VertexFormat::FLOAT4;
att.vertexAttrib = shaderinfos::VertexKey::VERTEX_ATTRIB_COLOR;
attribs.push_back(att);

auto mesh = Mesh::create(vertices, perVertexSizeInFloat, indices, attribs);

auto mat = MeshMaterial::createBuiltInMaterial(MeshMaterial::MaterialType::QUAD_COLOR, false);
mat->setPrimitiveType(backend::PrimitiveType::TRIANGLE_STRIP);
mesh->setMaterial(mat);

renderer->addMesh(mesh);

addChild(renderer);
```

**Reply by johnspeny** [reply]
I find no need for color if I just want to use the texture alone. Unfortunately the removal of color seams to crush the program producing an error. With color on my texture, appearance is being enhanced with color yet I am in need of just the original appearance of a texture.
![Undesk](https://user-images.githubusercontent.com/30299284/179922680-01cea3a0-34ce-4b41-b833-79f429748580.png)

Here is how I have written it.
```
	auto renderer = MeshRenderer::create();

	int perVertexSizeInFloat = 5;  // 3+2

	IndexArray indices;
	std::vector<float> vertices;
	indices.clear(CustomCommand::IndexFormat::U_INT);

	auto InsertVertex = [perVertexSizeInFloat, &vertices, &indices](float x, float y, float z, float u, float v)
	{
		unsigned int startindex = vertices.size() / perVertexSizeInFloat;
		vertices.insert(vertices.end(), {
			x,y,z, u,v,
			});

		indices.insert<uint32_t>(indices.size(),
			ilist_u32_t{ startindex });
	};

	InsertVertex(-100, -100, 0, 0, 0);
	InsertVertex(100, -100, 0, 1, 0);
	InsertVertex(0, 100, 0, .5, 1);

	std::vector<MeshVertexAttrib> attribs;
	MeshVertexAttrib att;

	att.type = backend::VertexFormat::FLOAT3;
	att.vertexAttrib = shaderinfos::VertexKey::VERTEX_ATTRIB_POSITION;
	attribs.push_back(att);

	att.type = backend::VertexFormat::FLOAT2;
	att.vertexAttrib = shaderinfos::VertexKey::VERTEX_ATTRIB_TEX_COORD;
	attribs.push_back(att);

	auto mesh = Mesh::create(vertices, perVertexSizeInFloat, indices, attribs);

	auto mat = MeshMaterial::createBuiltInMaterial(MeshMaterial::MaterialType::QUAD_TEXTURE, false);
	mat->setPrimitiveType(backend::PrimitiveType::TRIANGLE);

	mesh->setMaterial(mat);
	mesh->setTexture("HelloWorld.png");

	renderer->addMesh(mesh);

	addChild(renderer);
```

**Reply by DelinWorks** [reply]
you can just put all of the color values to 1.0 which is white, it would show the original texture this way,
or you can compile your own shader that only contains position and texture coordinate attributes and create a custom material with it! like this:
```
auto renderer = MeshRenderer::create();

int perVertexSizeInFloat = 5;  // 3+2

IndexArray indices;
std::vector<float> vertices;
indices.clear(CustomCommand::IndexFormat::U_INT);

auto InsertVertex = [perVertexSizeInFloat, &vertices, &indices](float x, float y, float z, float u, float v)
{
	unsigned int startindex = vertices.size() / perVertexSizeInFloat;
	vertices.insert(vertices.end(), {
		x,y,z,  u,v,
		});

	indices.insert<uint32_t>(indices.size(),
		ilist_u32_t{ startindex });
};

float offset = -100;

InsertVertex(-200 + offset * 2, -100 + offset, 0, 0, 1);
InsertVertex(200 + offset * 2, -100 + offset, 0, 1, 1);
InsertVertex(0 + offset * 2, 200 + offset, 0, .5, 0);

offset = 100;

InsertVertex(-200 + offset * 2, -100 + offset, 0, 0, 1);
InsertVertex(200 + offset * 2, -100 + offset, 0, 1, 1);
InsertVertex(0 + offset * 2, 200 + offset, 0, .5, 0);

std::vector<MeshVertexAttrib> attribs;
MeshVertexAttrib att;

att.type = backend::VertexFormat::FLOAT3;
att.vertexAttrib = shaderinfos::VertexKey::VERTEX_ATTRIB_POSITION;
attribs.push_back(att);

att.type = backend::VertexFormat::FLOAT2;
att.vertexAttrib = shaderinfos::VertexKey::VERTEX_ATTRIB_TEX_COORD;
attribs.push_back(att);

auto mesh = Mesh::create(vertices, perVertexSizeInFloat, indices, attribs);

const char* customVertShader = R"(
attribute vec4 a_position;
attribute vec2 a_texCoord;

varying vec2 TextureCoordOut;

uniform mat4 u_MVPMatrix;

void main()
{
    TextureCoordOut = a_texCoord;
    TextureCoordOut.y = 1.0 - TextureCoordOut.y;
    gl_Position = u_MVPMatrix * a_position;
}
)";

const char* customFragShader = R"(
#ifdef GL_ES
varying mediump vec2 TextureCoordOut;
varying mediump vec4 ColorOut;
#else
varying vec2 TextureCoordOut;
#endif
uniform vec4 u_color;

uniform sampler2D u_tex0;

void main(void)
{
    gl_FragColor = texture2D(u_tex0, TextureCoordOut) * u_color;
}
)";

auto shaderProgram = backend::Device::getInstance()->newProgram(customVertShader, customFragShader);
auto programState = new backend::ProgramState(shaderProgram);

auto mat = MeshMaterial::createWithProgramState(programState);
mat->setPrimitiveType(backend::PrimitiveType::TRIANGLE_STRIP);
mesh->setMaterial(mat);
mesh->setTexture("Untitled.png");

renderer->addMesh(mesh);

addChild(renderer);
```

Now it's rendering only the texture without any color values stored for each vertex:

![image](https://user-images.githubusercontent.com/45469625/180041639-09906816-2412-4ebc-b2e9-dfa0c2603b18.png)

**Reply by DelinWorks** [reply]
One advice is if you generate simple meshes (i.e. no more than 65000 vertices) then you should use unsigned short format (uint16_t) the code is simple:

```
IndexArray indices;
std::vector<float> vertices;
indices.clear(CustomCommand::IndexFormat::U_SHORT);

auto InsertVertex = [perVertexSizeInFloat, &vertices, &indices](float x, float y, float z, float u, float v)
{
	unsigned int startindex = vertices.size() / perVertexSizeInFloat;
	vertices.insert(vertices.end(), {
		x,y,z,  u,v,
		});

	indices.insert<uint16_t>(indices.size(),
		ilist_u16_t{ (uint16_t)startindex });
};
```

that would improve rendering performance especially on low-end devices.

**Reply by johnspeny** [reply]
> I was following a very old tutorial [here](https://chowdera.com/2021/09/20210902034412670v.html).

It says you can choose which points to be visible at a time for optimization process. Using this new approach since am creating triangles at once, I have now way of updating triangles to follow the updated points in view. Look at part 4 of the tutorial.

**Reply by johnspeny** [reply]
![failed](https://user-images.githubusercontent.com/30299284/182846144-9f458698-5ee7-4701-b892-30bc0387ff68.png)

**Reply by DelinWorks** [reply]
Sorry for the late response but you should be able to make a buffer with the size of maybe 12 vertices and then interpolate this vertex array based on screen size, you can do that as I mentioned above, you'd need to acquire the vertex buffer and update it just like this:

```
auto buffer = mesh->getMeshIndexData()->getVertexBuffer();
buffer->updateData(vertices.data(), vertices.size());
```

**Reply by johnspeny** [reply]
I tried to make a call to updateData every time the terrain moves offsets to the left, I get a memory access violation error on _meshIndexData. Then again all am trying to do is convert that old tutorial to the new approach. I also do not understand how and where to use the updateData(), its probably the last missing puzzle to get everything working well.

**Reply by DelinWorks** [reply]
use updateData after you've done creating the mesh. preferably in the update method

**Reply by johnspeny** [reply]
Tried that but nothing new, its harder than it sounds. Maybe my brain can't comprehend it well. There is something to do with the one creation of the mesh that maybe confusing me abit.I will probably share the public git repo and you have a glance.

**Reply by johnspeny** [reply]
Sorry for the delay. [The Repo](https://github.com/johnspeny/TerrainSample), I have tried to explain the flow, in case something is not clear don't hesitate to inform me.
